federation/pkg/federation-controller/util/federated_informer.go

```diff
@@ -127,13 +139,19 @@ func NewFederatedInformer(federationClient federation_release_1_4.Interface, tar
 			DeleteFunc: func(old interface{}) {
 				oldCluster, ok := old.(*federation_api.Cluster)
 				if ok {
+					if clusterLifecycle.ClusterUnavailable != nil {
+						clusterLifecycle.ClusterUnavailable(oldCluster, federatedInformer, federatedInformer.GetTargetStore())
+					}
```

Comment:
I'm wondering about the contract here - should we deliver "ClusterUnavailable" before deleting the cluster? If this will be asynchronous handler (which it basically should be for performance), the there will be a race with deleteCluster. This may cause a lot of pain.